xds/internal/balancer/cdsbalancer/cluster_handler_test.go

```diff
@@ -0,0 +1,609 @@
+	select {
+	case <-ch.updateChannel:
+		t.Fatal("Cluster Handler wrote an update to updateChannel when it shouldn't have, as each node in the full cluster tree has not yet received an update")
+	case <-shouldNotHappenCtx.Done():
+	}
+
+	// Invoke callback for Logical DNS child cluster.
+
+	fakeClient.InvokeWatchClusterCallback(xdsclient.ClusterUpdate{
+		ClusterType: xdsclient.ClusterTypeLogicalDNS,
+		ClusterName: logicalDNSService,
+	}, nil)
+
+	// Will Ping Cluster Handler, which will finally successfully build an update as all nodes in the tree of clusters
+	// have received an update.  Since this cluster is an aggregate cluster comprised of two children, the returned update
+	// should be length 2, as the xds cluster resolver LB policy only cares about the full list of LogicalDNS and EDS
+	// clusters representing the base nodes of the tree of clusters. This list should be ordered as per the cluster update.
+	select {
+	case chu := <-ch.updateChannel:
+		if diff := cmp.Diff(chu.chu, []xdsclient.ClusterUpdate{{
+			ClusterType: xdsclient.ClusterTypeEDS,
+			ClusterName: edsService,
+		}, {
+			ClusterType: xdsclient.ClusterTypeLogicalDNS,
+			ClusterName: logicalDNSService,
+		}}); diff != "" {
+			t.Fatalf("got unexpected cluster update, diff (-got, +want): %v", diff)
+		}
+	case <-ctx.Done():
+		t.Fatal("Timed out waiting for the cluster update to be written to the update buffer.")
+	}
+}
+
+// TestUpdateRootClusterAggregateThenChangeChild tests the scenario where you have an aggregate cluster with an EDS child
+// and a LogicalDNS child, then you change one of the children and send an update for the changed child. This should write
+// a new update to the update buffer to send back to CDS.
+func (s) TestUpdateRootClusterAggregateThenChangeChild(t *testing.T) {
+	// This initial code is the same as the test for the aggregate success case, except without validations. This will get
+	// this test to the point where it can change one of the children.
+	ch, fakeClient := setupTests(t)
+	ch.updateRootCluster(aggregateClusterService)
+
+	ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer ctxCancel()
+	_, err := fakeClient.WaitForWatchCluster(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
+	}
+
+	fakeClient.InvokeWatchClusterCallback(xdsclient.ClusterUpdate{
+		ClusterType:             xdsclient.ClusterTypeAggregate,
+		ClusterName:             aggregateClusterService,
+		PrioritizedClusterNames: []string{edsService, logicalDNSService},
+	}, nil)
+	fakeClient.InvokeWatchClusterCallback(xdsclient.ClusterUpdate{
+		ClusterType: xdsclient.ClusterTypeEDS,
+		ClusterName: edsService,
+	}, nil)
+	fakeClient.InvokeWatchClusterCallback(xdsclient.ClusterUpdate{
+		ClusterType: xdsclient.ClusterTypeLogicalDNS,
+		ClusterName: logicalDNSService,
+	}, nil)
+	fakeClient.WaitForWatchCluster(ctx)
```

Comment:
For readability, move these two `Wait()` right after `InvokeCallback(aggregated)`. As that's the order things would happen.

Reply:
Ah, good catch. Moved.